graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py

```diff
@@ -0,0 +1,1560 @@
+    ) -> None:
+        """Process and encode all graph data.
+
+        Extracts and encodes graph structure before writing to storage, then applies pre-processing
+        steps to node/edge features and labels, and saves transformed output to partitioned files,
+        one per node/edge type and feature.
+
+        As processing happens, each step returns values that we use to update
+        a common metadata dict that we eventually write to disk.
+
+        Parameters
+        ----------
+        data_configs : Mapping[str, Sequence[StructureConfig]]
+            Dictionary of configuration for nodes and edges
+        """
+        # TODO: See if it's better to return some data structure
+        # for the followup steps instead of just have side-effects
+        logging.info("Start Distributed Graph Processing ...")
+        process_start_time = perf_counter()
+
+        if not self._at_least_one_label_exists(data_configs):
+            self._insert_link_prediction_labels(data_configs["edges"])
+
+        metadata_dict = self._initialize_metadata_dict(data_configs)
+
+        edge_configs = data_configs["edges"]  # type: Sequence[EdgeConfig]
+
+        if "nodes" in data_configs:
+            node_configs = data_configs["nodes"]  # type: Sequence[NodeConfig]
+            missing_node_types = self._get_missing_node_types(edge_configs, node_configs)
+            if len(missing_node_types) > 0:
+                logging.info(
+                    "At least one node type missing (%s), creating node mapping from edges...",
+                    missing_node_types,
+                )
+                # In this case some node types only exist in the edge configs
+                id_map_start_time = perf_counter()
+                self.create_node_id_maps_from_edges(edge_configs, missing_node_types)
+                self.timers["create_node_id_maps_from_edges"] = perf_counter() - id_map_start_time
+
+            node_start_time = perf_counter()
+            metadata_dict["node_data"] = self.process_node_data(node_configs)
+            self.timers["process_node_data"] = perf_counter() - node_start_time
+        else:
+            # In this case no node files exist so we create all node mappings from the edge files
+            id_map_start_time = perf_counter()
+            missing_node_types = self._get_missing_node_types(edge_configs, [])
+            self.create_node_id_maps_from_edges(edge_configs, missing_node_types)
+            self.timers["create_node_id_maps_from_edges"] = perf_counter() - id_map_start_time
+            metadata_dict["node_data"] = {}
+
+        edges_start_time = perf_counter()
+        edge_data_dict, edges_dict = self.process_edge_data(edge_configs)
+        self.timers["process_edge_data"] = perf_counter() - edges_start_time
+        metadata_dict["edge_data"] = edge_data_dict
+        metadata_dict["edges"] = edges_dict
+        # We use the data location as the graph name, can also take from user?
```

Comment:
We could, I was trying to avoid adding another required user-facing argument here. Maybe I can add it as optional, and have this value when the user doesn't provide one?

Reply:
We can make it optional.

Reply:
OK this was done in the last commit, thanks.